File: client/src/types/subscription.ts
import { Category } from "./bill";

export type BillingFrequency = "MONTHLY" | "YEARLY" | "QUARTERLY" | "WEEKLY";
export type SubscriptionStatus = "ACTIVE" | "PAUSED" | "CANCELLED";

export interface Subscription {
  id: string;
  user_id: string;
  category_id?: string | null;
  service_name: string;
  plan_name?: string | null;
  cost: number;
  currency: string;
  billing_frequency: BillingFrequency;
  next_renewal_date: string; // YYYY-MM-DD
  status: SubscriptionStatus;
  trial_end_date?: string | null;
  cancellation_url?: string | null;
  created_at: string;
  updated_at: string;
  category?: Category | null;
}

export interface CreateSubscriptionPayload {
  service_name: string;
  plan_name?: string;
  cost: number;
  currency?: string;
  billing_frequency?: BillingFrequency;
  next_renewal_date: string;
  category_id?: string;
  trial_end_date?: string;
  cancellation_url?: string;
}

File: server/app/schemas/subscription.py
from datetime import date, datetime
from decimal import Decimal
from typing import Optional
from uuid import UUID
from pydantic import BaseModel, ConfigDict, Field
from app.schemas.category import CategoryResponse


class SubscriptionBase(BaseModel):
    service_name: str
    plan_name: Optional[str] = None
    cost: Decimal = Field(..., decimal_places=2)
    currency: str = "USD"
    billing_frequency: str = "MONTHLY"  # MONTHLY, YEARLY, QUARTERLY, WEEKLY
    next_renewal_date: date
    status: str = "ACTIVE"  # ACTIVE, PAUSED, CANCELLED
    trial_end_date: Optional[date] = None
    cancellation_url: Optional[str] = None


class SubscriptionCreate(SubscriptionBase):
    category_id: Optional[UUID] = None


class SubscriptionUpdate(BaseModel):
    service_name: Optional[str] = None
    plan_name: Optional[str] = None
    category_id: Optional[UUID] = None
    cost: Optional[Decimal] = None
    currency: Optional[str] = None
    billing_frequency: Optional[str] = None
    next_renewal_date: Optional[date] = None
    status: Optional[str] = None
    trial_end_date: Optional[date] = None
    cancellation_url: Optional[str] = None


class SubscriptionResponse(SubscriptionBase):
    id: UUID
    user_id: UUID
    category_id: Optional[UUID] = None
    created_at: datetime
    updated_at: datetime
    category: Optional[CategoryResponse] = None

    model_config = ConfigDict(from_attributes=True)
